### ISSUE TITLE
(Feature) Warn on potential data races

### ISSUE THREAD

**Issue by tbennun** [spcl/dace#424]
For example, a map that has the same transient name multiple times and the memlet ranges overlap should trigger a warning.

Might be expensive to check so may require adding a config entry for this.

**Comment by meshtag** [comment]
I am interested in working on this issue. Can someone provide a minimum example which produces an SDFG with memlet ranges overlap.

**Comment by tbennun** [comment]
@meshtag sounds good! In order to create this situation, one would have to manually create an SDFG with the API, since our Python frontend will not let the two statements be in one state.
There are many examples of how we create two-state SDFGs that have races here:
https://github.com/spcl/dace/blob/master/tests/transformations/state_fusion_test.py

Please let us know if you have a specific example you want to create or if we can help with the SDFG API!

**Comment by meshtag** [comment]
Apologies for the late reply.

I had a look at SDFG API and wanted to confirm whether the following SDFG is a valid example of memlet range overlap : 
```
import dace
import numpy as np

def memlet_range_overlap():
    sdfg = dace.SDFG('memlet_range_overlap')
    sdfg.add_array('A', [5], dace.int32)
    sdfg.add_array('B', [5], dace.int32)
    state = sdfg.add_state()
    r = state.add_read('A')
    t1 = state.add_tasklet('first_tasklet', {'a'}, {'b'}, 'b[0] = a[0] + 10')
    t2 = state.add_tasklet('second_tasklet', {'a'}, {'b'}, 'b[1] = a[1] - 20')
    w = state.add_write('B')
    state.add_edge(r, None, t1, 'a', dace.Memlet('A[0:2]'))
    state.add_edge(r, None, t2, 'a', dace.Memlet('A[0:2]'))
    state.add_edge(t1, 'b', w, None, dace.Memlet('B[0:2]'))
    state.add_edge(t2, 'b', w, None, dace.Memlet('B[2:4]'))

    A = np.ones([5], dtype=np.int32)
    B = np.zeros([5], dtype=np.int32)
    sdfg(A=A, B=B)
    print(A)
    print(B)

if __name__ == '__main__':
    memlet_range_overlap()

```

**Comment by tbennun** [comment]
Looks like it is not. The writes to B will be to indices 0,1 in t1; and 2,3 in t2.
There are overlapping reads but that should not be a data race issue.

**Comment by tbennun** [comment]
Merged in #1541, thanks!
